server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -2911,9 +2918,40 @@ public UserVm updateVirtualMachine(long id, String displayName, String group, Bo
             }
         }
 
+        if (!StringUtils.isEmpty(hostName)) {
+            vm.setHostName(hostName);
+            try {
+                updateDns(vm);
+            } catch (CloudRuntimeException e) {
+                throw new CloudRuntimeException(String.format("Failed to update hostname of VM %s to %s", vm.getInstanceName(), vm.getHostName()));
+            }
+        }
+
         return _vmDao.findById(id);
     }
 
+    private void updateDns(UserVmVO vm) {
+        List<NicVO> nicVOs = _nicDao.listByVmId(vm.getId());
+        for (NicVO nic : nicVOs) {
+            List<DomainRouterVO> routers = _routerDao.findByNetwork(nic.getNetworkId());
+            for (DomainRouterVO router : routers) {
+                Commands commands = new Commands(Command.OnError.Stop);
+                commandSetupHelper.createDhcpEntryCommand(router, vm, nic, false, commands);
+                try {
+                    _agentMgr.send(router.getHostId(), commands);
```

Comment:
@Pearl1594 can we use networkHelper.sendCommandsToRouter instead ?

https://github.com/apache/cloudstack/blob/f6073052aa3d592d6d737990a95c44e702908d59/server/src/main/java/com/cloud/network/router/NetworkHelperImpl.java#L180-L208

Reply:
Done @weizhouapache